### ISSUE TITLE
[Flang] Incorrect diagnostic on deallocating unlimited polymorphic pointer component

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#126806]
Consider the following code:
```
module m
    type base
        class (*), pointer :: data => null()         !! Error
        !real, pointer :: data => null()             !! Pass

        contains
        final :: finalizeBase
    end type

    contains

    elemental subroutine finalizeBase (b)
      type (base), intent(inout) :: b

      if (associated(b%data)) deallocate (b%data)
    end subroutine
end module
```

Flang is currently issue an error as
```
error: Semantic errors in t3.f
./t3.f:15:45: error: Object in DEALLOCATE statement is not deallocatable
        if (associated(b%data)) deallocate (b%data)
                                              ^^^^
./t3.f:15:45: because: 'data' is polymorphic in a pure subprogram
        if (associated(b%data)) deallocate (b%data)
                                              ^^^^
./t3.f:3:31: Declaration of 'data'
          class (*), pointer :: data => null()
                                ^^^^
```

The compilation is successful if the pointer component is intrinsic type.

**Comment by eugeneepshteyn** [comment]
It's interesting that ifx rejects this for the same reason:
```
$ ifx -c final-comp.f90 
final-comp.f90(15): error #5585: A statement that might result in the deallocation of a polymorphic entity is not permitted in a pure procedure.
      if (associated(b%data)) deallocate (b%data)
------------------------------^
compilation aborted for final-comp.f90 (code 1)
```
Based on this, I changed the example to remove finalization:
```
module m
    type base
        class (*), pointer :: data => null()         !! Error
        real, pointer :: data1 => null()             !! Pass
    end type

    contains

    pure subroutine destroyBase (b)
      type (base), intent(inout) :: b

      if (associated(b%data)) deallocate (b%data)
      if (associated(b%data1)) deallocate (b%data1)

    end subroutine
end module
```
And it fails similarly:
```
$ flang -c pure-comp.f90 
error: Semantic errors in pure-comp.f90
./pure-comp.f90:12:45: error: Object in DEALLOCATE statement is not deallocatable
        if (associated(b%data)) deallocate (b%data)
                                              ^^^^
./pure-comp.f90:12:45: because: 'data' is polymorphic in a pure subprogram
        if (associated(b%data)) deallocate (b%data)
                                              ^^^^
./pure-comp.f90:3:31: Declaration of 'data'
          class (*), pointer :: data => null()         !! Error
                                ^^^^
$ ifx -c pure-comp.f90 
pure-comp.f90(12): error #5585: A statement that might result in the deallocation of a polymorphic entity is not permitted in a pure procedure.
      if (associated(b%data)) deallocate (b%data)
```
Need to investigate, what's so special about  polymorphic entities in pure procedures.

**Comment by eugeneepshteyn** [comment]
Ok, found it, in section "15.7 Pure procedures": "C15106 A statement that might result in the deallocation of a polymorphic entity is not permitted in a pure procedure."

**Comment by DanielCChen** [comment]
> Ok, found it, in section "15.7 Pure procedures": "C15106 A statement that might result in the deallocation of a polymorphic entity is not permitted in a pure procedure."

Yes, you are right. I missed that. Thanks for checking it.
Flang compiled the original code successfully after I added `impure` to the finalization routine also.
